src/software/thunderscope/gl/layers/gl_draw_polygon_obstacle.py

```diff
@@ -0,0 +1,145 @@
+from software.thunderscope.gl.graphics.gl_polygon import GLPolygon
+from pyqtgraph.Qt import QtGui
+from pyqtgraph.Qt.QtCore import Qt
+from pyqtgraph.opengl import *
+
+from proto.import_all_protos import *
+from software.py_constants import *
+from software.thunderscope.gl.helpers.observable_list import ObservableList
+from software.thunderscope.proto_unix_io import ProtoUnixIO
+
+from software.thunderscope.gl.layers.gl_layer import GLLayer
+from software.thunderscope.gl.helpers.extended_gl_view_widget import MouseInSceneEvent
+
+
+class GLDrawPolygonObstacleLayer(GLLayer):
+    """A layer used to draw polygons that are going to represent obstacles for the trajectroy planner
+    to avoid.
+    """
+
+    def __init__(self, name: str, friendly_io: ProtoUnixIO) -> None:
+        """Initialize this layer
+
+        :param name:            the name of this layer
+        :param friendly_io:     the friendly_io
+        """
+        super().__init__(name)
+
+        self.friendly_io: ProtoUnixIO = friendly_io
+
+        self.current_polygon: GLPolygon = GLPolygon(parent_item=self, line_width=2)
+        # Tuple[float, float] represents a point (x,y)
+        self.points: List[Tuple[float, float]] = []
+
+        self.obstacles: List[Obstacle] = []
+
+        # used for keeping track and rendering multiple polygons
+        self.rendering_polygons: ObservableList = ObservableList(self._graphics_changed)
+
+    def keyPressEvent(self, event: QtGui.QKeyEvent) -> None:
+        """Responding to key events that are going to push obstacles to the stack or add point
+
+        :param event: a qt key event indicating the button that are pressed
+        """
+        if not self.visible():
+            return
+
+        if event.key() == Qt.Key.Key_P:
+            self.push_polygon_to_list()
+
+        if event.key() == Qt.Key.Key_C:
+            self.clear_polygons()
+
+    def clear_polygons(self):
+        """Clearing the obstacles"""
+        self.points.clear()
+        self.obstacles.clear()
+
+        for polygon in self.rendering_polygons:
+            polygon.hide()
+        self.rendering_polygons.resize(0, lambda: {})
+        self.current_polygon.hide()
+        self.current_polygon = GLPolygon(parent_item=self, line_width=2)
+
+        self._send_to_fs()
+
+    def push_polygon_to_list(self):
+        """Pushing the fully drawn polygon to the stack"""
+        points = [
+            Point(x_meters=point[0], y_meters=point[1]) for point in self.points[:-1]
+        ]
+
+        if len(points) <= 2:
+            print("Cannot push polygon to stack as there as less that points two.")
+            return
+
+        polygon = Polygon(points=points)
+        obstacle = Obstacle(polygon=polygon)
+        self.obstacles.append(obstacle)
+        self.points.clear()
+
+        self.rendering_polygons.append(self.current_polygon)
+        self.current_polygon = GLPolygon(parent_item=self, line_width=2)
+        self._send_to_fs()
```

Comment:
why do we create an obstacle here again and call `_send_to_fs`? by the time P is pressed, shouldn't the current polygon already be send to fullsystem through the `_add_one_point` method (which also calls `_send_to_fs`)? i feel like this method just needs to clear the current list of points

Reply:
this method would be more useful if the `_add_one_point` method is purely for visualization and doesn't actually send to fullsystem. that way we can finish drawing a polygon first, and then send to fullsystem at the end.

Reply:
i do like the `_add_one_point` sending to fullsystem dynamically tho so i think we can just make this method simpler